pkg/terraform/exec.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2021 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package terraform
+
+import (
+	"bytes"
+	"context"
+	"os"
+	"os/exec"
+)
+
+// Exec interface provides a Command interface. By this way, real or fake implementation of functions is determined.
+type Exec interface {
+	CommandContext(ctx context.Context, name string, dir string, env []string, arg ...string) Command
+}
+
+// RealExec struct implements the CommandContext function that calls the function from the exec package.
+type RealExec struct{}
+
+// CommandContext calls the function from exec package, then configure the *Cmd object with `dir` and `env`.
+// It returns a *Cmd object thus, the real implementation of CombinedOutput will be called.
+func (r *RealExec) CommandContext(ctx context.Context, name string, dir string, env []string, arg ...string) Command {
+	cmd := exec.CommandContext(ctx, name, arg...)
+
+	configureCmd(cmd, dir, env)
+
+	return cmd
+}
+
+// FakeExec struct implements the CommandContext function that calls fake one.
+type FakeExec struct {
+	stdErr []byte
```

Comment:
Do you mean that removing the stdErr and stdOut fields. I think you’re talking about manipulating these values by replacing them with functions. Like that:

```
type FakeExec struct {
	func... // instead of stdErr
	func... // instead of stdOut
	err    error
}
```

Reply:
This exec one can actually be used to initialize more than a single `Command` and users might want to return a different one depending on the arguments. For example, if the command is `terraform`, I can return a `Command` that returns a certain `stdOut` but if it's `cat`, I may want to return something else because these outputs are command-specific rather than exec library-specific. Giving users the ability to provide these differentiations would require a function context.

tldr; we wouldn't have any of those fields, but only a function to override what `CommandContext` does.